File: 3sem/4lab/Banks/Transactions/TransferTransaction.cs
using Banks.BankAccounts;
using Banks.Exceptions;
using Banks.Models;

namespace Banks.Transactions;

public class TransferTransaction : ITransaction
{
    public TransferTransaction(decimal sum, IBankAccount sender, IBankAccount receiver)
    {
        ArgumentNullException.ThrowIfNull(sender);
        ArgumentNullException.ThrowIfNull(receiver);

        if (sender.Id == receiver.Id)
            throw InvalidTransactionOperationException.OnTransactionToSameAccount();

        Id = Guid.NewGuid();
        Status = TransactionStatus.Created;

        Sum = sum;
        Sender = sender;
        Receiver = receiver;
    }

    public Guid Id { get; }
    public decimal Sum { get; }
    public IBankAccount Sender { get; }
    public IBankAccount Receiver { get; }
    public TransactionStatus Status { get; private set; }

    public void Execute()
    {
        if (Status is not TransactionStatus.Created)
            throw InvalidTransactionOperationException.OnIncorrectExecuting();

        if (!Sender.CanWithdraw(Sum) || !Receiver.CanDeposit(Sum))
            throw InvalidTransactionOperationException.OnInvalidTransaction();

        Sender.Withdraw(Sum);
        Receiver.Deposit(Sum);

        Status = TransactionStatus.Executed;
    }

    public void Cancel()
    {
        if (Status is not TransactionStatus.Executed)
            throw InvalidTransactionOperationException.OnIncorrectCanceling();

        if (!Receiver.CanWithdraw(Sum) || !Sender.CanDeposit(Sum))
            throw InvalidTransactionOperationException.OnInvalidTransaction();

        Receiver.Withdraw(Sum);
        Sender.Deposit(Sum);

        Status = TransactionStatus.Canceled;
    }
}

File: 4sem/1lab/src/main/java/BanksRemake/Domain/Transactions/TransferTransaction.java
package BanksRemake.Domain.Transactions;

import BanksRemake.Domain.BankAccounts.BankAccount;
import BanksRemake.Domain.Exceptions.InvalidBankAccountOperationException;
import BanksRemake.Domain.Exceptions.InvalidTransactionOperationException;
import BanksRemake.Domain.Models.TransactionStatus;

import java.util.UUID;

/**
 * The operation of transferring money from one account to another.
 */
public class TransferTransaction implements Transaction {
    private final UUID id;
    private final double sum;
    private final BankAccount sender;
    private final BankAccount receiver;
    private TransactionStatus status;

    public TransferTransaction(double sum, BankAccount sender, BankAccount receiver) throws InvalidTransactionOperationException {
        if (sender.getId() == receiver.getId()) {
            throw InvalidTransactionOperationException.onTransactionToSameAccount();
        }

        this.status = TransactionStatus.Created;

        this.sum = sum;
        this.sender = sender;
        this.receiver = receiver;

        this.id = UUID.randomUUID();
    }

    @Override
    public UUID getId() {
        return id;
    }

    @Override
    public double getSum() {
        return sum;
    }

    @Override
    public TransactionStatus getStatus() {
        return status;
    }

    @Override
    public void execute() throws InvalidTransactionOperationException, InvalidBankAccountOperationException {
        if (status != TransactionStatus.Created){
            throw InvalidTransactionOperationException.onIncorrectExecuting();
        }

        if (!sender.canWithdraw(sum) || !receiver.canDeposit(sum)) {
            throw InvalidTransactionOperationException.onInvalidTransaction();
        }

        sender.withdraw(sum);
        receiver.deposit(sum);

        status = TransactionStatus.Executed;
    }

    @Override
    public void cancel() throws InvalidTransactionOperationException, InvalidBankAccountOperationException {
        if (status != TransactionStatus.Executed){
            throw InvalidTransactionOperationException.onIncorrectCanceling();
        }

        if (!receiver.canWithdraw(sum) || !sender.canDeposit(sum)) {
            throw InvalidTransactionOperationException.onInvalidTransaction();
        }

        receiver.withdraw(sum);
        sender.deposit(sum);

        status = TransactionStatus.Canceled;
    }
}
